[SPARK-7855] Move bypassMergeSort-handling from ExternalSorter to own component

JoshRosen opened the pull request:
Spark's `ExternalSorter` writes shuffle output files during sort-based shuffle. Sort-shuffle contains a configuration, `spark.shuffle.sort.bypassMergeThreshold`, which causes ExternalSorter to skip sorting and merging and simply write separate files per partition, which are then concatenated together to form the final map output file.

The code paths used during this bypass are almost completely separate from ExternalSorter's other code paths, so refactoring them into a separate file can significantly simplify the code.

In addition to re-arranging code, this patch deletes a bunch of dead code.  The main entry point into ExternalSorter is `insertAll()` and in SPARK-4479 / #3422 this method was modified to completely bypass in-memory buffering of records when `bypassMergeSort` takes effect. As a result, some of the spilling and merging code paths will no longer be called when `bypassMergeSort` is used, so we should be able to safely remove that code.

There's an open JIRA ([SPARK-6026](https://issues.apache.org/jira/browse/SPARK-6026)) for removing the `bypassMergeThreshold` parameter and code paths; I have not done that here, but the changes in this patch will make removing that parameter significantly easier if we ever decide to do that.

This patch also makes several improvements to shuffle-related tests and adds more defensive checks to certain shuffle classes:
- DiskBlockObjectWriter now throws an exception if `fileSegment()` is called before `commitAndClose()` has been called.
- DiskBlockObjectWriter's close methods are now idempotent, so calling any of the close methods twice in a row will no longer result in incorrect shuffle write metrics changes.  Calling `revertPartialWritesAndClose()` on a closed DiskBlockObjectWriter now has no effect (before, it might mess up the metrics).
- The end-to-end shuffle record count metrics tests have been moved from InputOutputMetricsSuite to ShuffleSuite.  This means that these tests will now be run against all shuffle implementations rather than just the default shuffle configuration.
- The end-to-end metrics tests now include a test of a job which performs aggregation in the shuffle.
- Our tests now check that `shuffleBytesWritten == totalShuffleBytesRead`.
- FileSegment now throws IllegalArgumentException if it is constructed with a negative length or offset.

JoshRosen committed:
Move comparator methods closer together.

JoshRosen committed:
Remove duplicated buffer creation code.

JoshRosen committed:
Move diskBytesSpilled getter next to variable

JoshRosen committed:
WIP towards moving bypass code into own file.

JoshRosen committed:
Realize that bypass never buffers; proceed to delete tons of code

JoshRosen committed:
Add missing interface file

JoshRosen committed:
Delete more unused code

JoshRosen committed:
More simplification

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -123,43 +134,28 @@ private[spark] class ExternalSorter[K, V, C](
+  def diskBytesSpilled: Long = _diskBytesSpilled
 
-  // Write metrics for current spill
-  private var curWriteMetrics: ShuffleWriteMetrics = _
```
In the old code, it was pretty hard to trace through usages of this variable to figure out where it was reset / updated.  After removing the hash bypass code, I noticed that this variable's scope can be reduced to a local variable in `spillToMergebableFiles`.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -778,41 +692,15 @@ private[spark] class ExternalSorter[K, V, C](
 
     context.taskMetrics.incMemoryBytesSpilled(memoryBytesSpilled)
     context.taskMetrics.incDiskBytesSpilled(diskBytesSpilled)
-    context.taskMetrics.shuffleWriteMetrics.filter(_ => bypassMergeSort).foreach { m =>
```
Very happy to be removing this confusing block of code :)

JoshRosen replied:
To quickly clarify, this `filter` is equivalent to an `if(bypassMergeSort)` check.  I think that I haven't broken any of the metrics here for the non-bypass-merge-sort case, but it's a little tricky to tell because `ExternalSorterSuite` doesn't seem to have any tests that check the ShuffleWriteMetrics or task metrics.  I can look into adding these tests but I'd like to do it in a separate followup PR.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -778,41 +692,15 @@ private[spark] class ExternalSorter[K, V, C](
-    }
   }
 
-  def diskBytesSpilled: Long = _diskBytesSpilled
```
I moved this up to the top of the file so that it's next to where `_diskBytesSpilled` is defined.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -251,30 +243,21 @@ private[spark] class ExternalSorter[K, V, C](
-    } else {
-      spillToMergeableFile(collection)
-    }
+    spillToMergeableFile(collection)
```
I'll probably just merge this `spill` method into `spillToMergeableFile`.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33475/consoleFull) for   PR 6397 at commit [`02355ef`](https://github.com/apache/spark/commit/02355efd009ac2d667d49ab2f2e758447b463196).

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -238,11 +234,7 @@ private[spark] class ExternalSorter[K, V, C](
       }
     } else {
       if (maybeSpill(buffer, buffer.estimateSize())) {
-        buffer = if (useSerializedPairBuffer) {
```
This `if` statement for figuring out which buffer to use was replicated in two places, so I refactored it into a `newBuffer()` method, which is defined at the top of this file.  This allowed me to reduce the scope / visibility of some variables like `metaInitialRecords` and `kvChunkSize`.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -123,43 +134,28 @@ private[spark] class ExternalSorter[K, V, C](
   // grow internal data structures by growing + copying every time the number of objects doubles.
   private val serializerBatchSize = conf.getLong("spark.shuffle.spill.batchSize", 10000)
 
-  private def getPartition(key: K): Int = {
```
This was just moved higher up in the file so that it was next to the the other partitioner-related variables.

SparkQA commented:
[Test build #33475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33475/consoleFull) for   PR 6397 at commit [`02355ef`](https://github.com/apache/spark/commit/02355efd009ac2d667d49ab2f2e758447b463196).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33475/
Test FAILed.

JoshRosen committed:
Fix some style errors and comments.

JoshRosen committed:
Move conf prior to first use

JoshRosen committed:
Refactor logic for deciding when to bypass

JoshRosen committed:
Refactor tests.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33476 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33476/consoleFull) for   PR 6397 at commit [`931ca68`](https://github.com/apache/spark/commit/931ca686935f34c66964e7c26014c682716643f8).

JoshRosen committed:
Make sure to delete outputFile after errors.

SparkQA commented:
[Test build #33476 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33476/consoleFull) for   PR 6397 at commit [`931ca68`](https://github.com/apache/spark/commit/931ca686935f34c66964e7c26014c682716643f8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33476/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen committed:
Back out extra cleanup-handling code, since this is already covered in stop()

JoshRosen committed:
Fix style issue

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33478/consoleFull) for   PR 6397 at commit [`d267e0d`](https://github.com/apache/spark/commit/d267e0db9d85cf032d463e4b7992b25c82167a7c).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33477/
Test FAILed.

JoshRosen commented:
Woah, looks like the test log is being flooded with warnings about being unable to delete spill files:

```
13:26:02.102 ERROR org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter: Unable to delete file for partition i. 
13:26:02.102 ERROR org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter: Unable to delete file for partition i. 
13:26:02.102 ERROR org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter: Unable to delete file for partition i. 
13:26:02.102 ERROR org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter: Unable to delete file for partition i. 
```

It looks like I had the `if` condition negated; will fix now.

JoshRosen committed:
Add missing negation in deletion success check.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33478/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33480/consoleFull) for   PR 6397 at commit [`6320112`](https://github.com/apache/spark/commit/63201124b655c98208dc4d7da1d1afffe324ee64).

JoshRosen committed:
Combine spill() and spillToMergeableFile()

JoshRosen committed:
Make it more clear that curWriteMetrics is now only used for spill metrics

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33482/consoleFull) for   PR 6397 at commit [`0d9848c`](https://github.com/apache/spark/commit/0d9848cf3db0187b83c00ae5784ea358fa36f3b0).

JoshRosen commented:
Alright, I think that this is ready for review so I'm going to remove the `WIP` tag.  I'd like feedback on testing: I think that the tests here, combined with the implicit testing when this path is used during other shuffle write paths, should be sufficient to convince yourself that I haven't broken anything seriously when porting the bypass code into its own file.  It's worth calling out what I _don't_ have good unit tests for, though: we don't have any tests which explicitly try to read per-partition output from the combined output file.  AFAIK the `ExternalSorterSuite` didn't explicitly test this, either.  I might end up adding a separate test utility for emulating how IndexShuffleBlockResolver will read chunks of the output file and index file in order to serve shuffles.  This is part of an effort to avoid writing new end-to-end tests for testing basic shuffle functionality, since those tend to be slow and aren't great for testing things like metrics.

JoshRosen commented:
/cc @rxin @pwendell, I think this is a good first step towards cleaning up ExternalSorter and would love to get your feedback on this.  This is a stepping-stone to simplify this code before some planned refactorings to the DiskObjectWriter interface.

SparkQA commented:
**[Test build #33480 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33480/consoleFull)**     for PR 6397 at commit [`6320112`](https://github.com/apache/spark/commit/63201124b655c98208dc4d7da1d1afffe324ee64)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33480/
Test FAILed.

SparkQA commented:
[Test build #33482 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33482/consoleFull) for   PR 6397 at commit [`0d9848c`](https://github.com/apache/spark/commit/0d9848cf3db0187b83c00ae5784ea358fa36f3b0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33482/
Test FAILed.

JoshRosen committed:
Rename SortShuffleSorter to SortShuffleFileWriter

This is a better name, since one of the implementations never sorts.

JoshRosen committed:
Remove unnecessary overloaded methods

JoshRosen committed:
Fix bug when aggregator is used but map-side combine is disabled

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33494/consoleFull) for   PR 6397 at commit [`bc1a820`](https://github.com/apache/spark/commit/bc1a8205fd0f1254966dcfbb337ef2ee05d83186).

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+  /** Array of file writers for each partition */
+  private[this] var partitionWriters: Array[BlockObjectWriter] = _
+
+  def insertAll(records: Iterator[_ <: Product2[K, V]]): Unit = {
```
override

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+   * @param context a TaskContext for a running Spark task, for us to update shuffle metrics.
+   * @return array of lengths, in bytes, of each partition of the file (used by map output tracker)
+   */
+  def writePartitionedFile(blockId: BlockId, context: TaskContext, file: File): Array[Long] = {
```
override

SparkQA commented:
[Test build #33494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33494/consoleFull) for   PR 6397 at commit [`bc1a820`](https://github.com/apache/spark/commit/bc1a8205fd0f1254966dcfbb337ef2ee05d83186).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33494/
Test FAILed.

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+  def writePartitionedFile(blockId: BlockId, context: TaskContext, file: File): Array[Long] = {
+    if (partitionWriters == null) {
+      // We were passed an empty iterator
+      Array.fill(numPartitions)(0L)
```
i think i'd prefer an explicit return here. i was lookin at the code and thought : "oh what's going on. isn't this a no-op".

JoshRosen replied:
I was purposely trying to avoid ever using `return` in Scala code.  I can add it back if you think it will improve clarity.

rxin replied:
btw this is also just
```scala
new Array[Long](numPartitions)
```

and much faster this way too

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+      val out = new FileOutputStream(file, true)
+      val writeStartTime = System.nanoTime
+      Utils.tryWithSafeFinally {
+        for (i <- 0 until numPartitions) {
```
a while loop maybe? this is fairly expensive for a large number of partitions

JoshRosen replied:
I think that this was a carryover from the old code, but I don't mind changing it here.

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+  def stop(): Unit = {
+    if (partitionWriters != null) {
+      partitionWriters.foreach { w =>
+        w.revertPartialWritesAndClose()
```
can this ever throw an exception? if yes we should silent it.

JoshRosen replied:
Meh, I'll just rewrite this file in Java so that we have checked exceptions...

JoshRosen replied:
Actually, this method explicitly doesn't throw exceptions.  From its Scaladoc:

```
  /**
   * Reverts writes that haven't been flushed yet. Callers should invoke this function
   * when there are runtime exceptions. This method will not throw, though it may be
   * unsuccessful in truncating written data.
   */
  def revertPartialWritesAndClose()
```

rxin commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -84,35 +86,40 @@ import org.apache.spark.storage.{BlockObjectWriter, BlockId}
+
+  // Since SPARK-7855, bypassMergeSort optimization is no longer performed as part of this class.
+  // As a sanity check, make sure that we're not handling a shuffle which should use that path.
+  if (SortShuffleWriter.shouldBypassMergeSort(conf, numPartitions, aggregator, ordering)) {
```
maybe turn this into an assert with an error msg?

JoshRosen replied:
This is called once per map task, so there's not an obvious performance benefit to allowing users to disable assertions to skip this check.

My intent was for this to guard against misuse of this ExternalSorter for cases that should have been handled by the bypassMergeSort path.  The idea is that the non-sort-shuffle uses of this sort will pass an ordering and thus will not be eligible for the bypass, so any sort case which has no aggregation, no key ordering, and few partitions shouldn't have gone through this path if it had few partitions.  One corner-case is situations where we haven't specified an ordering, partitioner, or aggregator; in this case, I think that `ExternalSorter` will effectively do nothing, so therefore we should treat that configuration as an error.  I'll look into adding some error-checking to detect this case.

rxin commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -123,43 +130,28 @@ private[spark] class ExternalSorter[K, V, C](
-    new PartitionedSerializedPairBuffer[K, C](metaInitialRecords, kvChunkSize, serInstance)
-  } else {
-    new PartitionedPairBuffer[K, C]
+  private def newBuffer: () => WritablePartitionedPairCollection[K, C] with SizeTracker = {
```
why not just return WritablePartitionedPairCollection?

JoshRosen replied:
Ah, I got a bit too clever here.  My idea was to reduce the visibility of variables like `useSerializedPairBuffer` without having to recompute them on each call to `newBuffer(0)`, so I had this block of code effectively return a factory for creating buffers based on a configuration which is read once.

rxin commented on core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala:
```diff
@@ -47,13 +47,20 @@ private[spark] trait WritablePartitionedPairCollection[K, V] {
-  }
+    val it = partitionedDestructiveSortedIterator(keyComparator)
+    new WritablePartitionedIterator {
+      var cur = if (it.hasNext) it.next() else null
```
private[this]?

JoshRosen replied:
Yep; this was a carryover from the old code, but we should clean it up.

rxin commented on core/src/main/scala/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.scala:
```diff
@@ -0,0 +1,132 @@
+import org.apache.spark.util.collection._
+
+/**
+ * This class handles sort-based shuffle's `bypassMergeSort` write path, which is used for shuffles
```
one thing missing from this comment is what does bypassmergesort actually means.

JoshRosen replied:
I'll rewrite this Scaladoc to provide a better explanation and to also address how this is different than HashShuffleWriter (the key difference is the fact that this output in a format that can be served with IndexShuffleBlockResolver).

rxin commented:
@zsxwing please review this one as well. Thanks!

JoshRosen commented:
By the way, I know why the test is failing: I overlooked the fact that we don't perform map-side key sorting unless we're also combining by key.  I'll restructure the code to make this more obvious.

JoshRosen committed:
Convert BypassMergeSortShuffleWriter to Java

JoshRosen committed:
Add missing private[this]

JoshRosen committed:
Add comment to clarify confusing factory code

JoshRosen committed:
Update Javadoc

JoshRosen commented:
I've addressed the review comments.  I also ported BypassMergeSortShuffleWriter to Java and revised its Javadoc comment to try to better explain how it works.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33564/consoleFull) for   PR 6397 at commit [`71d76ff`](https://github.com/apache/spark/commit/71d76ff6c882a010f85168b3db7770fa5f1808d2).

SparkQA commented:
[Test build #33564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33564/consoleFull) for   PR 6397 at commit [`71d76ff`](https://github.com/apache/spark/commit/71d76ff6c882a010f85168b3db7770fa5f1808d2).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33564/
Test FAILed.

zsxwing commented on core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala:
```diff
@@ -49,18 +50,27 @@ private[spark] class SortShuffleWriter[K, V, C](
         dep.aggregator, Some(dep.partitioner), dep.keyOrdering, dep.serializer)
-      sorter.insertAll(records)
+    } else if (SortShuffleWriter.shouldBypassMergeSort(
+        SparkEnv.get.conf, dep.partitioner.numPartitions, aggregator = None, dep.keyOrdering)) {
```
In the previous codes, if `dep.mapSideCombine == false`, the ordering is `None`. Why do you change it to `dep.keyOrdering`?

JoshRosen replied:
This was a mistake that I forgot to fix in my last push (the tests catch it though); will fix this morning.

Sent from my phone

> On May 27, 2015, at 7:38 AM, Shixiong Zhu <notifications@github.com> wrote:
> 
> In core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala:
> 
> >          dep.aggregator, Some(dep.partitioner), dep.keyOrdering, dep.serializer)
> > -      sorter.insertAll(records)
> > +    } else if (SortShuffleWriter.shouldBypassMergeSort(
> > +        SparkEnv.get.conf, dep.partitioner.numPartitions, aggregator = None, dep.keyOrdering)) {
> In the previous codes, if dep.mapSideCombine == false, the ordering is None. Why do you change it to dep.keyOrdering?
> 
> —
> Reply to this email directly or view it on GitHub.
>

zsxwing commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -123,43 +130,30 @@ private[spark] class ExternalSorter[K, V, C](
-    if (shouldPartition) partitioner.get.getPartition(key) else 0
+  // The idea here is to reduce minimize the visibility of variables like `kvChunkSize` and
+  // `useSerializedPairBuffer` without recomputing them on each call to `newBuffer()`.
+  private def newBuffer: () => WritablePartitionedPairCollection[K, C] with SizeTracker = {
```
`newBuffer` should be `val`, or it will recompute each time.

JoshRosen replied:
I'll just roll this back to a simpler version that does not attempt to hide those variables inside of the function, since this type of excessively clever code does not belong in the shuffle.

Sent from my phone

> On May 27, 2015, at 8:05 AM, Shixiong Zhu <notifications@github.com> wrote:
> 
> In core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
> 
> > @@ -123,43 +130,30 @@ private[spark] class ExternalSorter[K, V, C](
> >    // grow internal data structures by growing + copying every time the number of objects doubles.
> >    private val serializerBatchSize = conf.getLong("spark.shuffle.spill.batchSize", 10000)
> >  
> > -  private def getPartition(key: K): Int = {
> > -    if (shouldPartition) partitioner.get.getPartition(key) else 0
> > +  // The idea here is to reduce minimize the visibility of variables like `kvChunkSize` and
> > +  // `useSerializedPairBuffer` without recomputing them on each call to `newBuffer()`.
> > +  private def newBuffer: () => WritablePartitionedPairCollection[K, C] with SizeTracker = {
> newBuffer should be val, or it will recompute each time.
> 
> —
> Reply to this email directly or view it on GitHub.
>

JoshRosen committed:
Add missing overrides; fix compilation

JoshRosen committed:
Remove excessively clever (and wrong) implementation of newBuffer()

JoshRosen committed:
Do not perform a map-side sort unless we're also doing map-side aggregation

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33597/consoleFull) for   PR 6397 at commit [`8522b6a`](https://github.com/apache/spark/commit/8522b6a033e5bdb8b07c6f24973f646da6ac6a3c).

SparkQA commented:
[Test build #33597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33597/consoleFull) for   PR 6397 at commit [`8522b6a`](https://github.com/apache/spark/commit/8522b6a033e5bdb8b07c6f24973f646da6ac6a3c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33597/
Test FAILed.

JoshRosen commented:
Hmm, interesting test failure:

```
org.apache.spark.metrics.InputOutputMetricsSuite.shuffle records written metrics
 Error Details
0 did not equal 100000
```

JoshRosen committed:
Remove confusing taskMetrics.shuffleWriteMetrics() optional call

JoshRosen commented:
Ah, I found the problem: we're closing the partition writers at the wrong place.  We should be calling `commitAndClose()` at the end of `insertAll()`.  The current usage of DiskBlockObjectWriter isn't legal because we call `fileSegment()` before we've called `commitAndClose()`.  

According to BlockObjectWriter's API contract:

```scala
  /**
   * Returns the file segment of committed data that this Writer has written.
   * This is only valid after commitAndClose() has been called.
   */
  def fileSegment(): FileSegment
```

However, DiskBlockObjectWriter doesn't have assertions to catch if `fileSegment()` is called at the incorrect time.  I'll add these assertions, then fix my new writer code.

I also noticed that InputOutputMetricsSuite is one of the only places with an end-to-end test of the shuffle records written metric.  This test probably belongs in ShuffleSuite, since it should be tested for each shuffle manager.  We should also add another test to cover these metrics for shuffles which perform aggregation, since these shuffles may use a different code path.

JoshRosen committed:
Guard against calling fileSegment() before commitAndClose() has been called.

JoshRosen commented:
Actually, I spoke slightly too soon: it looks like we _do_ call `commitAndClose()` in `writePartitionedFile()`, but then it looks like we turn around and call `revertPartialWritesAndClose()` in `stop()` on a DiskBlockObjectWriter that's already closed.  It looks like this ends up rolling back the shuffle metrics for the block that we've already committed.  We probably shouldn't be calling `revertPartialWritesAndClose()` here anyways, but I think that BlockObjectWriter should guard against calling revert on a writer that's already been closed by making that operation into a no-op if the writer is already closed.

JoshRosen committed:
Add more tests + defensive programming to DiskBlockObjectWriter.

JoshRosen committed:
Move shuffle metrics tests to ShuffleSuite.

JoshRosen committed:
Minor fix to cleanup logic.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen commented:
I added a bunch more tests to DiskBlockObjectWriter to ensure that all of the close methods are idempotent.  I've also moved the metrics test from InputOutputMetricsSuite into ShuffleSuite so that they're run against all shuffle managers and have added a test to check that the metrics are correct when shuffles perform aggregation.  I also added some additional checks to ensure that shuffleBytesRead == shuffleBytesWritten.  /cc @ksakellis, since this touches some of your code in InputOutputMetricsSuite.

SparkQA commented:
[Test build #33681 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33681/consoleFull) for   PR 6397 at commit [`03f35a4`](https://github.com/apache/spark/commit/03f35a476025516084c2b8a072abf97d1a91bc3f).

JoshRosen commented:
In case reviewers missed this in the updated PR description, here's a better summary of the new changes:

This patch also makes several improvements to shuffle-related tests and adds more defensive checks to certain shuffle classes:

- DiskBlockObjectWriter now throws an exception if `fileSegment()` is called before `commitAndClose()` has been called.
- DiskBlockObjectWriter's close methods are now idempotent, so calling any of the close methods twice in a row will no longer result in incorrect shuffle write metrics changes.  Calling `revertPartialWritesAndClose()` on a closed DiskBlockObjectWriter now has no effect (before, it might mess up the metrics).
- The end-to-end shuffle record count metrics tests have been moved from InputOutputMetricsSuite to ShuffleSuite.  This means that these tests will now be run against all shuffle implementations rather than just the default shuffle configuration.
- The end-to-end metrics tests now include a test of a job which performs aggregation in the shuffle.
- Our tests now check that `shuffleBytesWritten == totalShuffleBytesRead`.

JoshRosen commented:
@sryza, this might also be of interest to you.

SparkQA commented:
[Test build #33681 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33681/consoleFull) for   PR 6397 at commit [`03f35a4`](https://github.com/apache/spark/commit/03f35a476025516084c2b8a072abf97d1a91bc3f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen committed:
Guard against negative offsets and lengths in FileSegment

JoshRosen commented:
Hmm, looks like this fails nearly all of the HashShuffleSuite tests.  I found the problem: after 8b8fb9e3bba2944ae1967a9a532ab39ff19a7deb, the `finalPosition` in DiskBlockObjectWriter was no longer being set when calling `commitAndClose()` on writers that hadn't ever been opened.  This caused `fileSegment()` to return a segment with a negative length, causing problems during shuffle fetching / serving.  I've fixed this problem and added `require`s to `FileSegment` to guard against negative offsets and lengths.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33716 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33716/consoleFull) for   PR 6397 at commit [`8b216c4`](https://github.com/apache/spark/commit/8b216c4593dc39f49cb9b1abdaae1a720a183866).

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33723/consoleFull) for   PR 6397 at commit [`8b216c4`](https://github.com/apache/spark/commit/8b216c4593dc39f49cb9b1abdaae1a720a183866).

SparkQA commented:
[Test build #33716 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33716/consoleFull) for   PR 6397 at commit [`8b216c4`](https://github.com/apache/spark/commit/8b216c4593dc39f49cb9b1abdaae1a720a183866).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #33723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33723/consoleFull) for   PR 6397 at commit [`8b216c4`](https://github.com/apache/spark/commit/8b216c4593dc39f49cb9b1abdaae1a720a183866).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

zsxwing commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -0,0 +1,184 @@
+    while (records.hasNext()) {
+      final Product2<K, V> record = records.next();
+      final K key = record._1();
+      partitionWriters[partitioner.getPartition(key)].write(key, record._2());
```
`partitioner.getPartition(key)` is different from the previous codes. The previous codes  calls `getPartition` which doesn't call `partitioner.getPartition(key)` if there is only 1 partition. I'm not sure if such optimization does matter.

JoshRosen replied:
Thanks for noticing this.  I left this out on purpose, but should have probably commented on the diff to explain why.

I think that the reason why `ExternalSorter` skips the `partitioner.getPartition(key)` call when there is only one partition is because `ExternalSorter` is also used for non-shuffle contexts for which we don't define a partitioner (such as the reduce-side sort in `sortByKey()`.  In those cases, we obviously want to avoid unnecessary hashing.

BypassMergeSortShuffleWriter is only used for shuffles, though, and I expect that it's extremely rare to have shuffles that shuffle everything to a single partition (collecting results to the driver is handled by different code).  Therefore, I chose to leave out that check.

zsxwing replied:
I see. Fair enough. LGTM now.

zsxwing commented:
LGTM except the `partitioner.getPartition(key)` issue.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into external-sorter-bypass-cleanup

Conflicts:
	core/src/test/scala/org/apache/spark/storage/BlockObjectWriterSuite.scala
	core/src/test/scala/org/apache/spark/util/collection/ExternalSorterSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33775/consoleFull) for   PR 6397 at commit [`bf3f3f6`](https://github.com/apache/spark/commit/bf3f3f6bab69cae01056fd21e2f6d238e0ab707d).

SparkQA commented:
[Test build #33775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33775/consoleFull) for   PR 6397 at commit [`bf3f3f6`](https://github.com/apache/spark/commit/bf3f3f6bab69cae01056fd21e2f6d238e0ab707d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -238,46 +230,33 @@ private[spark] class ExternalSorter[K, V, C](
-   * Spill the current in-memory collection to disk, adding a new file to spills, and clear it.
-   */
-  override protected[this] def spill(collection: WritablePartitionedPairCollection[K, C]): Unit = {
-    if (bypassMergeSort) {
```
To clarify a little further, since we no longer buffered records in memory when `bypassMergeSort` was true, this branch would never be taken.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -724,28 +653,7 @@ private[spark] class ExternalSorter[K, V, C](
-    if (bypassMergeSort && partitionWriters != null) {
-      // We decided to write separate files for each partition, so just concatenate them. To keep
-      // this simple we spill out the current in-memory collection so that everything is in files.
-      spillToPartitionFiles(if (aggregator.isDefined) map else buffer)
```
I think that this call was effectively a no-op since `buffer` should now be empty when `bypassMergeSort = true`.

JoshRosen commented on core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala:
```diff
@@ -778,41 +686,15 @@ private[spark] class ExternalSorter[K, V, C](
-  /**
-   * Read a partition file back as an iterator (used in our iterator method)
-   */
-  private def readPartitionFile(writer: BlockObjectWriter): Iterator[Product2[K, C]] = {
```
This was only called from one place, `partitionedIterator`, when `bypassMergeSort` is true.  `partitionedIterator`, in turn, is called when merging multiple spills.  Since `bypassMergeSort` no longer spills, then `partitionedIterator` would never be called and its `bypassMergeSort` merging code is no longer necessary.

rxin commented:
LGTM.
